### PR TITLE
Implement detailed lifetime control for querynode

### LINKED ISSUE #21846: [Bug]: Data race in querynode lifetime control
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 09124f8
- Deployment mode(standalone or cluster): both
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

Data race in waitgroup usage(add after wait) in querynode lifetime control

### Expected Behavior

No data race

### Steps To Reproduce

_No response_

### Milvus Log

```
==================
2023-01-28T12:03:40.1921787Z WARNING: DATA RACE
2023-01-28T12:03:40.1925635Z Write at 0x00c000cf0320 by goroutine 36:
2023-01-28T12:03:40.1925925Z   runtime.racewrite()
2023-01-28T12:03:40.1926183Z       <autogenerated>:1 +0x24
2023-01-28T12:03:40.1926684Z   github.com/milvus-io/milvus/internal/querynode.(*QueryNode).Stop.func1()
2023-01-28T12:03:40.1927188Z       /go/src/github.com/milvus-io/milvus/internal/querynode/query_node.go:358 +0x784
2023-01-28T12:03:40.1927516Z   sync.(*Once).doSlow()
2023-01-28T12:03:40.1927784Z       /usr/local/go/src/sync/once.go:68 +0x101
2023-01-28T12:03:40.1928043Z   sync.(*Once).Do()
2023-01-28T12:03:40.1928299Z       /usr/local/go/src/sync/once.go:59 +0x46
2023-01-28T12:03:40.1928732Z   github.com/milvus-io/milvus/internal/querynode.(*QueryNode).Stop()
2023-01-28T12:03:40.1929219Z       /go/src/github.com/milvus-io/milvus/internal/querynode/query_node.go:335 +0x64
2023-01-28T12:03:40.1929705Z   github.com/milvus-io/milvus/internal/distributed/querynode.(*Server).Stop()
2023-01-28T12:03:40.1930237Z       /go/src/github.com/milvus-io/milvus/internal/distributed/querynode/service.go:219 +0x18e
2023-01-28T12:03:40.1930724Z   github.com/milvus-io/milvus/internal/proxy.TestProxy.func5()
2023-01-28T12:03:40.1931208Z       /go/src/github.com/milvus-io/milvus/internal/proxy/proxy_test.go:439 +0x39
2023-01-28T12:03:40.1931546Z   runtime.deferreturn()
2023-01-28T12:03:40.1931835Z       /usr/local/go/src/runtime/panic.go:436 +0x32
2023-01-28T12:03:40.1932102Z   testing.tRunner()
2023-01-28T12:03:40.1932378Z       /usr/local/go/src/testing/testing.go:1439 +0x213
2023-01-28T12:03:40.1932651Z   testing.(*T).Run.func1()
2023-01-28T12:03:40.1932932Z       /usr/local/go/src/testing/testing.go:1486 +0x47
2023-01-28T12:03:40.1933109Z 
2023-01-28T12:03:40.1933250Z Previous read at 0x00c000cf0320 by goroutine 413:
2023-01-28T12:03:40.1933524Z   runtime.raceread()
2023-01-28T12:03:40.1933773Z       <autogenerated>:1 +0x24
2023-01-28T12:03:40.1934186Z   github.com/milvus-io/milvus/internal/querynode.(*QueryNode).GetMetrics()
2023-01-28T12:03:40.1934859Z       /go/src/github.com/milvus-io/milvus/internal/querynode/impl.go:1280 +0xf8
2023-01-28T12:03:40.1935372Z   github.com/milvus-io/milvus/internal/distributed/querynode.(*Server).GetMetrics()
2023-01-28T12:03:40.1935908Z       /go/src/github.com/milvus-io/milvus/internal/distributed/querynode/service.go:318 +0x76
2023-01-28T12:03:40.1936575Z   github.com/milvus-io/milvus/internal/proto/querypb._QueryNode_GetMetrics_Handler.func1()
2023-01-28T12:03:40.1937116Z       /go/src/github.com/milvus-io/milvus/internal/proto/querypb/query_coord.pb.go:5215 +0x9e
2023-01-28T12:03:40.1937649Z   github.com/milvus-io/milvus/internal/util/logutil.UnaryTraceLoggerInterceptor()
2023-01-28T12:03:40.1938200Z       /go/src/github.com/milvus-io/milvus/internal/util/logutil/grpc_interceptor.go:22 +0x71
2023-01-28T12:03:40.1938746Z   github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1()
2023-01-28T12:03:40.1939316Z       /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x8e
2023-01-28T12:03:40.1939828Z   go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1()
2023-01-28T12:03:40.1940377Z       /go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.37.0/interceptor.go:349 +0x6da
2023-01-28T12:03:40.1940973Z   github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1()
2023-01-28T12:03:40.1941535Z       /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x8e
2023-01-28T12:03:40.1942066Z   github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1()
2023-01-28T12:03:40.1942627Z       /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34 +0x126
2023-01-28T12:03:40.1943144Z   github.com/milvus-io/milvus/internal/proto/querypb._QueryNode_GetMetrics_Handler()
2023-01-28T12:03:40.1943682Z       /go/src/github.com/milvus-io/milvus/internal/proto/querypb/query_coord.pb.go:5217 +0x22c
2023-01-28T12:03:40.1944066Z   google.golang.org/grpc.(*Server).processUnaryRPC()
2023-01-28T12:03:40.1944428Z       /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:1340 +0x1589
2023-01-28T12:03:40.1944769Z   google.golang.org/grpc.(*Server).handleStream()
2023-01-28T12:03:40.1946059Z       /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:1713 +0xfd3
2023-01-28T12:03:40.1946425Z   google.golang.org/grpc.(*Server).serveStreams.func1.2()
2023-01-28T12:03:40.1946793Z       /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:965 +0xec
2023-01-28T12:03:40.1946992Z 
2023-01-28T12:03:40.1947114Z Goroutine 36 (running) created at:
2023-01-28T12:03:40.1947367Z   testing.(*T).Run()
2023-01-28T12:03:40.1947647Z       /usr/local/go/src/testing/testing.go:1486 +0x724
2023-01-28T12:03:40.1947948Z   testing.runTests.func1()
2023-01-28T12:03:40.1948238Z       /usr/local/go/src/testing/testing.go:1839 +0x99
2023-01-28T12:03:40.1948509Z   testing.tRunner()
2023-01-28T12:03:40.1948779Z       /usr/local/go/src/testing/testing.go:1439 +0x213
2023-01-28T12:03:40.1949049Z   testing.runTests()
2023-01-28T12:03:40.1949329Z       /usr/local/go/src/testing/testing.go:1837 +0x7e4
2023-01-28T12:03:40.1949594Z   testing.(*M).Run()
2023-01-28T12:03:40.1949868Z       /usr/local/go/src/testing/testing.go:1719 +0xa71
2023-01-28T12:03:40.1950277Z   github.com/milvus-io/milvus/internal/proxy.TestMain()
2023-01-28T12:03:40.1950751Z       /go/src/github.com/milvus-io/milvus/internal/proxy/task_index_test.go:40 +0x33
2023-01-28T12:03:40.1951059Z   main.main()
2023-01-28T12:03:40.1951293Z       _testmain.go:2113 +0x40f
2023-01-28T12:03:40.1951445Z 
2023-01-28T12:03:40.1951564Z Goroutine 413 (running) created at:
2023-01-28T12:03:40.1951879Z   google.golang.org/grpc.(*Server).serveStreams.func1()
2023-01-28T12:03:40.1952238Z       /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:963 +0x4bd
2023-01-28T12:03:40.1952640Z   google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
2023-01-28T12:03:40.1953102Z       /go/pkg/mod/google.golang.org/grpc@v1.51.0/internal/transport/http2_server.go:599 +0x2fa6
2023-01-28T12:03:40.1953633Z   google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
2023-01-28T12:03:40.1954054Z       /go/pkg/mod/google.golang.org/grpc@v1.51.0/internal/transport/http2_server.go:644 +0x2eb
2023-01-28T12:03:40.1954428Z   google.golang.org/grpc.(*Server).serveStreams()
2023-01-28T12:03:40.1954834Z       /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:949 +0x266
2023-01-28T12:03:40.1955192Z   google.golang.org/grpc.(*Server).handleRawConn.func1()
2023-01-28T12:03:40.1955549Z       /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:891 +0x64
2023-01-28T12:03:40.1955843Z ==================
```

### Anything else?

https://github.com/milvus-io/milvus/actions/runs/4031583663/jobs/6930957227

### PR DESCRIPTION
**Pull request by congqixia**
Resolves #21846 

/kind bug

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

### CODE REVIEW

File: internal/util/lifetime/lifetime.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// GetState returns current state.
+func (l *lifetime[T]) GetState() T {
+	l.mut.RLock()
+	defer l.mut.RUnlock()
+
+	return l.state
+}
+
+// Add records a task is running, returns false if the lifetime is not healthy.
+func (l *lifetime[T]) Add(isHealthy IsHealthy[T]) bool {
```
Comment (weiliu1031) [inline internal/util/lifetime/lifetime.go:80]: `Add` seems confused here, consider to change it like `Check` or `Statisfy` ?

File: internal/querynode/impl.go [hunk 2]
```diff
@@ -47,13 +47,11 @@ import (
 )
 
 // isHealthy checks if QueryNode is healthy
-func (node *QueryNode) isHealthy() bool {
-	code := node.stateCode.Load().(commonpb.StateCode)
+func (node *QueryNode) isHealthy(code commonpb.StateCode) bool {
```
Comment (weiliu1031) [inline internal/querynode/impl.go:50]: `isHealthy` and `isHealthyOrStopping` method seems no need to bind with querynode instance?

File: internal/util/lifetime/lifetime.go (hunk 1 above)
Comment (congqixia) [reply internal/util/lifetime/lifetime.go:80]: `Add` actually do call the same method on internal wait group, and shall be paired with `Done` method. Naming is kind of difficult to determine here.

File: internal/querynode/impl.go (hunk 2 above)
Comment (congqixia) [reply internal/querynode/impl.go:50]: True. Will more them into `common` package if any other component is reusing `Lifetime` component.

**Pull request merged**